### PR TITLE
ci: check if Cargo.lock is up-to-date

### LINKED ISSUE #11992: Check if a pull request forgot to commit Cargo.lock in CI
### Problem

Starting from #11851, this project commits `Cargo.lock` into git history. However, CI has yet got a good process to detect the error when a change of `Cargo.lock` should be committed but not. This kind of failure is now caught by `docs` job, which effectively runs `ci/validate-man.sh`. 

For example, CI failures below detect `Cargo.lock` changed and was not committed.

- https://github.com/rust-lang/cargo/actions/runs/4731403990/jobs/8396280788
- https://github.com/rust-lang/cargo/actions/runs/4731845144/jobs/8397266389

### Proposed Solution

We shouldn't rely on `ci/validate-man.sh`, as it serves for a different purpose. We may add a new CI job or step to detect if `Cargo.lock` changes and git status is dirty after a build.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Fixes #11992. I believe there is no need to check git status here.

85dc569ac132d8fa369a9a6678542f4dc63f45ed from #11840 is also cherry-picked, since it doesn't need to check changes for the entire git repository.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -97,6 +97,8 @@ jobs:
       run: echo CARGO_CONTAINER_TESTS=1 >> $GITHUB_ENV
       if: matrix.os == 'ubuntu-latest'
 
+    - name: Check if Cargo.lock is up-to-date
+      run: cargo fetch --locked
```
Comment (epage) [inline .github/workflows/main.yml:101]: The downside with doing this in a tet job is that it means someone doesn't collect any feedback if this error occurs.  This is also why distinct clippy jobs are nice so warnings about builds don't prevent seeing test errors or the other way around

Comment (weihanglo) [reply .github/workflows/main.yml]: That's fair. Moved it to an independent job. I think it's fine keeping things going for now, but maintainers should keep an eye on lockfile being updated.

As a result of being a separate job, it uses more resource and energy, which is not really great 😆.

Comment (epage) [reply .github/workflows/main.yml]: Being a 13s job, I'm not too worried.  If this was more in the minutes, then I would be.

**Pull request merged**